File: projects/13-live-ingest/src/live.rs
//! The live registry + per-stream window — the shared state between the RTMP
//! producers (publisher sessions) and the HTTP consumers (viewers).
//!
//! This is **plumbing** — it is (mostly) implemented. Each live stream is a bounded
//! ring of built segments/parts plus a `watch` channel that publishes the current
//! **live edge** `(media-sequence, part)`; producers push built fMP4 bytes and bump
//! the edge, consumers read the window and (for LL-HLS blocking reload, V4) *await*
//! the edge. What lands here is already-built `Bytes` — the *building* (V3, `fmp4.rs`)
//! and the *playlist rendering* + blocking policy (V4, `llhls.rs`) are the `todo!()`s;
//! this module just holds and hands out the pieces, memoized so N viewers share one
//! mux (the fan-out contract).
//!
//! A live stream is unbounded in time, so the window is a *fixed* size — RAM tracks
//! `window_segments`, not airtime. That is the whole reason a 10-hour broadcast fits
//! in a few seconds of memory.

use std::collections::{HashMap, VecDeque};
use std::sync::{Arc, Mutex};

use bytes::Bytes;
use chrono::{DateTime, Utc};
use tokio::sync::watch;

/// Immutable ingest configuration, shared (behind an `Arc`) by every session and
/// handler.
#[derive(Debug, Clone)]
pub struct IngestConfig {
    /// Authorized stream keys. Empty ⇒ allow any key (dev only — see `authorize`).
    pub stream_keys: Vec<String>,
    /// LL-HLS part target in seconds (~200–350 ms).
    pub target_part_secs: f64,
    /// Full-segment target in seconds.
    pub target_segment_secs: f64,
    /// How many finished segments to retain in the live window (the memory bound).
    pub window_segments: usize,
}

/// The live edge: the newest `(media_sequence, part)` a consumer can ask for. LL-HLS
/// blocking reload (V4) waits until this reaches a requested `(msn, part)`.
///
/// Ordered `msn`-major then `part`-minor so `await_edge` is a simple `>=` compare.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord)]
pub struct LiveEdge {
    pub msn: u64,
    pub part: u64,
}

/// One built partial segment (an LL-HLS **part**, ~200 ms of fMP4).
#[derive(Debug, Clone)]
pub struct Part {
    pub bytes: Bytes,
    pub duration: f64,
    /// True when this part begins on an IDR keyframe (`#EXT-X-PART:…,INDEPENDENT=YES`).
    pub independent: bool,
}

/// One media segment in the live window: a media sequence number, its parts, and —
/// once closed — the concatenated full-segment bytes.
#[derive(Debug, Clone)]
pub struct Segment {
    pub msn: u64,
    pub parts: Vec<Part>,
    /// `false` while still accumulating parts at the live edge.
    pub complete: bool,
    /// Full-segment bytes, present once `complete` (init + all parts is what a
    /// non-LL player fetches).
    pub bytes: Option<Bytes>,
    pub duration: f64,
    /// Wall-clock start, for `#EXT-X-PROGRAM-DATE-TIME`.
    pub program_date_time: DateTime<Utc>,
}

/// One live stream's bounded window of built media plus its edge notifier.
pub struct LiveStream {
    inner: Mutex<StreamInner>,
    edge_tx: watch::Sender<LiveEdge>,
    window_segments: usize,
}

struct StreamInner {
    /// CMAF init segment (`ftyp`+`moov`), built once by V3 when the codec config lands.
    init: Option<Bytes>,
    /// Rolling window of segments; the front is the oldest retained, the back is the
    /// one currently forming at the live edge.
    segments: VecDeque<Segment>,
    /// Media sequence number of the next segment to open (monotonic; never reused).
    next_msn: u64,
    /// True once the publisher has disconnected — the playlist gets `#EXT-X-ENDLIST`.
    ended: bool,
}

impl LiveStream {
    fn new(window_segments: usize) -> Self {
        let (edge_tx, _) = watch::channel(LiveEdge::default());
        Self {
            inner: Mutex::new(StreamInner {
                init: None,
                segments: VecDeque::new(),
                next_msn: 0,
                ended: false,
            }),
            edge_tx,
            window_segments,
        }
    }

    // -- producer side (called from a publisher session, V3) -----------------

    /// Install the CMAF init segment (once, when the codec config is known).
    pub fn set_init(&self, init: Bytes) {
        self.inner.lock().unwrap().init = Some(init);
    }

    /// Append a freshly built part to the segment currently forming at the live edge,
    /// opening a new segment first if `start_segment` is true (a keyframe boundary).
    /// Bumps the live edge, unparking any held blocking-reload requests (V4).
    pub fn push_part(&self, part: Part, start_segment: bool) {
        let edge = {
            let mut g = self.inner.lock().unwrap();
            if start_segment || g.segments.is_empty() {
                let msn = g.next_msn;
                g.next_msn += 1;
                g.segments.push_back(Segment {
                    msn,
                    parts: Vec::new(),
                    complete: false,
                    bytes: None,
                    duration: 0.0,
                    program_date_time: Utc::now(),
                });
                // Trim the window from the front to keep memory bounded.
                while g.segments.len() > self.window_segments {
                    g.segments.pop_front();
                }
            }
            let seg = g.segments.back_mut().expect("segment just ensured");
            seg.duration += part.duration;
            seg.parts.push(part);
            LiveEdge {
                msn: seg.msn,
                part: (seg.parts.len() - 1) as u64,
            }
        };
        // Ignore send errors: no receivers just means no one is blocking right now.
        let _ = self.edge_tx.send(edge);
    }

    /// Close the segment currently at the live edge, recording its full-segment bytes
    /// (init-independent concatenation of its parts, built by V3).
    pub fn finish_segment(&self, full_bytes: Bytes) {
        let mut g = self.inner.lock().unwrap();
        if let Some(seg) = g.segments.back_mut() {
            seg.complete = true;
            seg.bytes = Some(full_bytes);
        }
    }

    /// Mark the stream ended (publisher gone) — the playlist gains `#EXT-X-ENDLIST`.
    pub fn mark_ended(&self) {
        self.inner.lock().unwrap().ended = true;
    }

    // -- consumer side (called from HTTP handlers, V4) -----------------------

    /// The current live edge.
    pub fn edge(&self) -> LiveEdge {
        *self.edge_tx.borrow()
    }

    /// Await the live edge reaching `target` (LL-HLS blocking reload). Returns the
    /// edge once `edge() >= target`, or the last edge if the stream ends first. The
    /// *policy* (what a blocking `_HLS_msn/_HLS_part` maps to, the timeout) is V4 in
    /// `llhls.rs`; this is just the park-until-signalled mechanism it builds on.
    pub async fn await_edge(&self, target: LiveEdge) -> LiveEdge {
        let mut rx = self.edge_tx.subscribe();
        loop {
            let cur = *rx.borrow_and_update();
            if cur >= target {
                return cur;
            }
            if rx.changed().await.is_err() {
                return cur; // producer gone
            }
        }
    }

    pub fn is_ended(&self) -> bool {
        self.inner.lock().unwrap().ended
    }

    /// The init segment, if built yet.
    pub fn init_bytes(&self) -> Option<Bytes> {
        self.inner.lock().unwrap().init.clone()
    }

    /// The full bytes of a completed segment by media sequence number.
    pub fn segment_bytes(&self, msn: u64) -> Option<Bytes> {
        let g = self.inner.lock().unwrap();
        g.segments
            .iter()
            .find(|s| s.msn == msn)
            .and_then(|s| s.bytes.clone())
    }

    /// The bytes of one part `(msn, part)` if still in the window.
    pub fn part_bytes(&self, msn: u64, part: u64) -> Option<Bytes> {
        let g = self.inner.lock().unwrap();
        g.segments
            .iter()
            .find(|s| s.msn == msn)
            .and_then(|s| s.parts.get(part as usize))
            .map(|p| p.bytes.clone())
    }

    /// A cloned snapshot of the window, for rendering the playlist (V4). Cloning is
    /// cheap: `Bytes` clones are refcount bumps, not copies.
    pub fn snapshot(&self) -> (Vec<Segment>, bool) {
        let g = self.inner.lock().unwrap();
        (g.segments.iter().cloned().collect(), g.ended)
    }
}

/// The set of live streams, keyed by stream key. Publisher sessions create entries;
/// viewers look them up.
pub struct LiveRegistry {
    streams: Mutex<HashMap<String, Arc<LiveStream>>>,
    cfg: Arc<IngestConfig>,
}

impl LiveRegistry {
    pub fn new(cfg: Arc<IngestConfig>) -> Self {
        Self {
            streams: Mutex::new(HashMap::new()),
            cfg,
        }
    }

    pub fn config(&self) -> &IngestConfig {
        &self.cfg
    }

    /// Is this key allowed to publish? An empty allow-list means "any key" (dev).
    ///
    /// TODO(security, horizontal): this is the auth gate the SPEC grades. A static
    /// allow-list is the floor; a real deployment verifies a signed token / calls an
    /// auth service here. Never log the key itself — log a hash/prefix.
    pub fn authorize(&self, key: &str) -> bool {
        self.cfg.stream_keys.is_empty() || self.cfg.stream_keys.iter().any(|k| k == key)
    }

    /// Get-or-create the live stream for a key (publisher side).
    pub fn open(&self, key: &str) -> Arc<LiveStream> {
        let mut g = self.streams.lock().unwrap();
        g.entry(key.to_string())
            .or_insert_with(|| Arc::new(LiveStream::new(self.cfg.window_segments)))
            .clone()
    }

    /// Look up a live stream for a key (viewer side); `None` if nothing is on air.
    pub fn get(&self, key: &str) -> Option<Arc<LiveStream>> {
        self.streams.lock().unwrap().get(key).cloned()
    }

    /// Remove a stream when its publisher disconnects.
    pub fn close(&self, key: &str) {
        self.streams.lock().unwrap().remove(key);
    }

    /// Keys currently on air, sorted — for `GET /live`.
    pub fn live_keys(&self) -> Vec<String> {
        let mut keys: Vec<String> = self.streams.lock().unwrap().keys().cloned().collect();
        keys.sort();
        keys
    }
}

File: projects/13-live-ingest/src/live_ingest/live.py
"""The live registry and per-stream window — **wired** plumbing, not a vertical.

This is the shared state between the two planes: publisher sessions (RTMP)
write built fMP4 bytes into it, viewers (HTTP) read them back out. Each live
stream is a bounded ring of segments and parts plus a **live edge** — the newest
`(msn, part)` a viewer can ask for — and a signal that fires whenever that edge
moves, which is what a blocking playlist reload parks on.

What lands here is already-built `bytes`. The *building* is V3 (`fmp4.py`) and
the *playlist rendering* is V4 (`llhls.py`); this module only holds the pieces
and hands them out.

## Build once, serve N — and why `bytes` makes that free

A part is muxed once, on the publisher's session, and stored here. Two hundred
viewers fetching it get two hundred references to **the same `bytes` object**:
Python's `bytes` is immutable, so sharing one is a refcount increment, never a
copy — the direct equivalent of cloning a refcounted `Bytes` in Rust. Starlette
writes a `Response(content=...)` body straight from that object. Keep it that
way: a `bytearray` here, or a `bytes(view)` per request, quietly turns fan-out
into per-viewer memcpy, and the boss fight's "each part muxed once" counter will
not catch it because the *mux* still happened once.

## Bounded memory on an unbounded stream

A broadcast never ends on its own, so the window is a fixed ring:
`collections.deque(maxlen=window_segments)`. Appending a new segment to a full
deque silently drops the oldest from the front — the eviction *is* the append.
RAM tracks the window, not airtime, which is why a ten-hour stream fits in the
same few megabytes as a ten-minute one.

## Why there are no locks

The Rust version put the window behind a `Mutex` because sessions and handlers
ran on a multi-threaded runtime. Here everything runs on **one event loop
thread**, and none of these methods `await` — so each runs to completion before
any other coroutine gets the thread, which makes each one atomic with respect to
every session and every handler. That guarantee holds exactly as long as nobody
adds an `await` in the middle of a mutation. If you ever do, you have written a
race, and no test on one loop will show it to you.

## Parking many requests on one signal

`await_edge` is the mechanism LL-HLS blocking reload needs: hundreds of held
playlist requests that all wake on the next `push_part`, with no thread and no
poll loop per request. It uses the *generation event* idiom: each edge change
sets the current `asyncio.Event` — waking every coroutine parked on it at once —
and installs a fresh one for the next change. A waiter re-checks its target
after waking, so a wake that does not reach its target just parks it again. The
*policy* built on top (what a request waits for, and for how long) is V4's.
"""

from __future__ import annotations

import asyncio
import dataclasses
from collections import deque
from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import NamedTuple

from .config import Settings

__all__ = [
    "LiveEdge",
    "LiveRegistry",
    "LiveStream",
    "Part",
    "Segment",
    "WindowSnapshot",
]


class LiveEdge(NamedTuple):
    """The newest `(msn, part)` a viewer can ask for.

    A `NamedTuple` for one reason: tuples compare lexicographically, so
    `LiveEdge(12, 3) >= LiveEdge(12, 1)` and `LiveEdge(13, 0) > LiveEdge(12, 9)`
    are already the right ordering — msn major, part minor — with no `__lt__`
    to write or get wrong.
    """

    msn: int
    part: int


@dataclass(frozen=True, slots=True)
class Part:
    """One built LL-HLS part: ~200–350 ms of fMP4 (`moof` + `mdat`)."""

    data: bytes
    duration: float
    """Seconds — what `#EXT-X-PART:DURATION=` advertises."""
    independent: bool
    """True when this part begins on an IDR keyframe (`INDEPENDENT=YES`)."""


@dataclass(slots=True)
class Segment:
    """One media segment in the window: its msn, its parts, and — once closed —
    the full-segment bytes a non-LL player fetches."""

    msn: int
    program_date_time: datetime
    """Wall clock at the segment's start, for `#EXT-X-PROGRAM-DATE-TIME`."""
    parts: list[Part] = field(default_factory=list[Part])
    complete: bool = False
    """False while still accumulating parts at the live edge."""
    data: bytes | None = None
    """Present once `complete`."""
    duration: float = 0.0


@dataclass(frozen=True, slots=True)
class WindowSnapshot:
    """What the playlist renderer (V4) walks: the window at one instant.

    The `segments` are copies whose `parts` lists are copied too, so a renderer
    holding a snapshot never sees a part appear mid-render. That copy is a list
    of references — the part *bytes* are never copied.
    """

    segments: tuple[Segment, ...]
    ended: bool
    edge: LiveEdge | None


class LiveStream:
    """One live stream's bounded window of built media, plus its edge signal."""

    def __init__(self, key: str, window_segments: int) -> None:
        self.key = key
        self._init: bytes | None = None
        self._segments: deque[Segment] = deque(maxlen=window_segments)
        self._next_msn = 0
        """Monotonic and never reused: an msn permanently names its bytes, which
        is what makes `Cache-Control: immutable` on a segment honest."""
        self._ended = False
        self._edge: LiveEdge | None = None
        self._edge_changed = asyncio.Event()

    # -- producer side (a publisher session, V2/V3) -------------------------

    def set_init(self, data: bytes) -> None:
        """Install the CMAF init segment, once the codec config is known."""
        self._init = data

    def push_part(self, part: Part, *, start_segment: bool) -> LiveEdge:
        """Append a built part at the live edge and wake every parked waiter.

        `start_segment=True` (a keyframe boundary) opens a new segment first;
        so does the very first part. Opening a segment on a full window evicts
        the oldest one. Returns the new edge.
        """
        if start_segment or not self._segments:
            self._segments.append(Segment(msn=self._next_msn, program_date_time=datetime.now(UTC)))
            self._next_msn += 1
        segment = self._segments[-1]
        segment.parts.append(part)
        segment.duration += part.duration
        self._edge = LiveEdge(segment.msn, len(segment.parts) - 1)
        self._wake()
        return self._edge

    def finish_segment(self, data: bytes) -> None:
        """Close the segment at the live edge with its full-segment bytes."""
        if self._segments:
            segment = self._segments[-1]
            segment.complete = True
            segment.data = data

    def mark_ended(self) -> None:
        """The publisher is gone: the playlist gains `#EXT-X-ENDLIST`, and every
        held request wakes rather than waiting on an edge that will never move."""
        self._ended = True
        self._wake()

    def _wake(self) -> None:
        self._edge_changed.set()
        self._edge_changed = asyncio.Event()

    # -- consumer side (HTTP handlers, V4) ----------------------------------

    @property
    def edge(self) -> LiveEdge | None:
        """The current live edge, or `None` before the first part exists."""
        return self._edge

    @property
    def ended(self) -> bool:
        return self._ended

    @property
    def init(self) -> bytes | None:
        """The init segment, or `None` until V3 has built it."""
        return self._init

    async def await_edge(self, target: LiveEdge) -> LiveEdge | None:
        """Park until the edge reaches `target`, or the stream ends.

        Returns the edge at that moment. No timeout here on purpose — the bound
        is policy, and policy is V4's (`llhls.media_playlist` wraps this in one).
        """
        while True:
            if self._edge is not None and self._edge >= target:
                return self._edge
            if self._ended:
                return self._edge
            # Read the event into a local *before* awaiting: `_wake` replaces
            # the attribute, and a waiter must park on the generation that was
            # current when it checked, or it can miss the wake entirely.
            changed = self._edge_changed
            await changed.wait()

    def segment_bytes(self, msn: int) -> bytes | None:
        """Full bytes of a *completed* segment still in the window."""
        segment = self._find(msn)
        return segment.data if segment is not None and segment.complete else None

    def part_bytes(self, msn: int, part: int) -> bytes | None:
        """Bytes of part `(msn, part)`, if still in the window."""
        segment = self._find(msn)
        if segment is None or not 0 <= part < len(segment.parts):
            return None
        return segment.parts[part].data

    def snapshot(self) -> WindowSnapshot:
        """The window at this instant, safe to render from. See `WindowSnapshot`."""
        return WindowSnapshot(
            segments=tuple(
                dataclasses.replace(segment, parts=list(segment.parts))
                for segment in self._segments
            ),
            ended=self._ended,
            edge=self._edge,
        )

    def _find(self, msn: int) -> Segment | None:
        # msns in the window are contiguous, so this is index arithmetic rather
        # than a scan — O(1) however large the window is configured.
        if not self._segments:
            return None
        index = msn - self._segments[0].msn
        if 0 <= index < len(self._segments):
            return self._segments[index]
        return None


class LiveRegistry:
    """The streams currently on air, keyed by stream key.

    Publisher sessions create entries; viewers look them up. A plain `dict` —
    see the module docstring on why a single event loop needs no lock around it.
    """

    def __init__(self, settings: Settings) -> None:
        self._streams: dict[str, LiveStream] = {}
        self._allowed = settings.allowed_keys
        self._window_segments = settings.live_window_segments

    def authorize(self, key: str) -> bool:
        """May this key publish? An empty allow-list means any key (dev only).

        TODO(security, horizontal): this is the auth gate the SPEC grades. A
        static allow-list is the floor; a real deployment verifies a signed
        token or calls an `on_publish` webhook here. Two things worth deciding
        deliberately: plain `in` short-circuits on the first differing byte, so
        compare secrets with `hmac.compare_digest`; and never log `key` — log a
        hash or a short prefix.
        """
        return not self._allowed or key in self._allowed

    def open(self, key: str) -> LiveStream:
        """Get-or-create the stream for `key` (publisher side).

        TODO(V2, security): get-or-create means a *second* publisher to a key
        that is already live gets the same window and interleaves its parts
        with the first one's. Whether that is refused, or takes over, is a
        decision the publish state machine has to make before calling this.
        """
        stream = self._streams.get(key)
        if stream is None:
            stream = LiveStream(key, self._window_segments)
            self._streams[key] = stream
        return stream

    def get(self, key: str) -> LiveStream | None:
        """The stream for `key` (viewer side), or `None` if nothing is on air."""
        return self._streams.get(key)

    def close(self, key: str) -> None:
        """Remove a stream when its publisher disconnects.

        Viewers already holding a reference keep it; `mark_ended` (called first)
        is what wakes them.
        """
        self._streams.pop(key, None)

    def live_keys(self) -> list[str]:
        """Keys currently on air, sorted — for `GET /live`."""
        return sorted(self._streams)

    def __len__(self) -> int:
        return len(self._streams)
